### PR TITLE
v2 [WIP]

### PR DESCRIPTION
**Pull request by Salakar**
## Change Log

Version 2.0.0 of RNFirebase introduces a new features, breaking changes, deprecations ([matching v4 of the web SDK](https://firebase.google.com/support/release-notes/js#4.0.0)) and general build / dependency improvements.

### Ads

- [android] Implemented Admob Banner view
- [both] Implemented Admob Interstitials
- [both] Implemented Admob Rewarded Video
- [android] Implemented Admob Native Express

### Performance Monitoring
RNFirebase now supports the recently released [Performance Monitoring](https://firebase.google.com/docs/perf-mon/) module. Much like Crash Reporting, automatic events (traces) are delivered to your console whilst providing support to register custom traces. See the [documentation](http://invertase.io/react-native-firebase/#/modules/perf) on how to use this in your app.

**Implemented methods:**
- setPerformanceCollectionEnabled(enabled: `boolean`)
- newTrace(id: `string`): `Trace`
- `Trace`.start(): `void`
- `Trace`.incrementCounter(event: `string`): `void`
- `Trace`.stop(): `void`

### Remote Config
- [both] Implemented remote config in [v1.1.0](https://github.com/invertase/react-native-firebase/releases/tag/v1.1.0) - adding here for sake of completeness. [Documentation](http://invertase.io/react-native-firebase/#/modules/config)

### Auth

- [both] [fetchProvidersForEmail()](https://firebase.google.com/docs/reference/js/firebase.auth.Auth#fetchProvidersForEmail): `Promise<Array<String>>` implemented

### Crash

- [android] Implemented new crash log disabling methods.

## Bug Fixes
- [js] Fixed an issue whereby the result of `onAuthStateChange` returned a different result on state changes.

## Breaking Changes & Deprecations
- [breaking] **RNFirebase v2.0.0 drops support for React Native versions less than v40**
- [deprecated] `providerId` should now be used instead of `provider` whilst obtaining auth credentials. The latter will be removed in future releases.
- [deprecated] Deprecated `User.getToken` in favour of `User.getIdToken`.
- [breaking] `User.reauthenticate` has been removed in favour of `User.reauthenticateWithCredential`.
- [breaking] `User.link ` has been removed in favour of `User.linkWithCredential`.
- [breaking] Removed unnecessary `didReceiveNotificationResponse` and `willPresentNotification` methods for iOS messaging. Added additional `didReceiveRemoteNotification` method.
- [breaking] firebase.messaging().onTokenRefresh and firebase.messaging().onMessage return a function to unsubscribe as per the Web SDK spec: https://firebase.google.com/docs/reference/js/firebase.messaging.Messaging#onMessage.  Previously they returned an object with a .remove() method.

### Android
All firebase modules are now optional so you only need to import the Firebase functionality that you require in your application.

You need to make a couple of changes to your `app/build.gradle` file.  Update the react-native-firebase compile statement to read:

```gradle
compile(project(':react-native-firebase')) {
  transitive = false
}
compile "com.google.firebase:firebase-core:11.0.0"
```
Add each of the firebase modules you need from the following list:
```gradle
compile "com.google.firebase:firebase-ads:11.0.0"
compile "com.google.firebase:firebase-analytics:11.0.0"
compile "com.google.firebase:firebase-auth:11.0.0"
compile "com.google.firebase:firebase-config:11.0.0"
compile "com.google.firebase:firebase-crash:11.0.0"
compile "com.google.firebase:firebase-database:11.0.0"
compile "com.google.firebase:firebase-messaging:11.0.0"
compile "com.google.firebase:firebase-perf:11.0.0"
compile "com.google.firebase:firebase-storage:11.0.0"
```
Update `MainApplication.java` and import the modules as required:
```java
import io.invertase.firebase.admob.RNFirebaseAdMobPackage; //Firebase AdMob
import io.invertase.firebase.analytics.RNFirebaseAnalyticsPackage; // Firebase Analytics
import io.invertase.firebase.auth.RNFirebaseAuthPackage; // Firebase Auth
import io.invertase.firebase.config.RNFirebaseRemoteConfigPackage; // Firebase Remote Config
import io.invertase.firebase.crash.RNFirebaseCrashPackage; // Firebase Crash Reporting
import io.invertase.firebase.database.RNFirebaseDatabasePackage; // Firebase Realtime Database
import io.invertase.firebase.messaging.RNFirebaseMessagingPackage; // Firebase Cloud Messaging
import io.invertase.firebase.perf.RNFirebasePerformancePackage; // Firebase Performance Monitoring
import io.invertase.firebase.storage.RNFirebaseStoragePackage; // Firebase Storage
```
Add the packages to the `getPackages()` method as required:
```java
@Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
          new MainReactPackage(),
          new RNFirebasePackage(),  // <-- Add this line
          // Add these packages as appropriate
          new RNFirebaseAdMobPackage(),
          new RNFirebaseAnalyticsPackage(),
          new RNFirebaseAuthPackage(),
          new RNFirebaseRemoteConfigPackage(),
          new RNFirebaseCrashPackage(),
          new RNFirebaseDatabasePackage(),
          new RNFirebaseMessagingPackage(),
          new RNFirebasePerformancePackage(),
          new RNFirebaseStoragePackage()
      );
    }
```

### iOS
All firebase modules are now optional so you only need to import the Firebase functionality that you require in your application.  Simply update your Podfile to only include the Firebase modules for functionality that you require in your app.

### CODE REVIEW

**Comment by Salakar** [comment]
cc @Ehesp @chrisbianca

**Comment by sraka1** [comment]
@Salakar what is the ETA for this being production-ready/merged to master?

Also, I noticed "confirmPasswordReset" and "applyActionCode" methods are missing on Auth. Should I open an issue related to v2 or is it already in the plans?

**Comment by Ehesp** [comment]
@sraka1 I'll take a look at those tomorrow.

**Comment by Salakar** [comment]
@sraka1 can't give you a specific ETA as we've still got quite a bit to do and this is our first big release so we don't want to rush it

In the meantime you could always use v2 branch as your module source on npm - we do try not to push up any broken/untested code (but it can still happen) so generally you're safe to use it whilst developing your app and switch it out for the official release when it's ready - if you're desperate. 😸

File: android/build.gradle [hunk 1]
```diff
@@ -2,35 +2,36 @@
 // YES, jcenter is required twice - it somehow tricks studio into compiling deps below
 // doesn't break anything anywhere else and projects using this lib work as normal
 buildscript {
-    ext.firebaseVersion = '10.2.6'
+    ext.firebaseVersion = '11.0.0'
```
Comment (HofmannZ) [inline android/build.gradle]: Uses version 11.0.0 here

File: docs/installation-android.md [hunk 2]
```diff
@@ -39,18 +52,43 @@ public class MainApplication extends Application implements ReactApplication {
...
 
-```java
+```
 dependencies {
-  compile project(':react-native-firebase')
+  # RNFirebase Required dependencies
+  compile(project(':react-native-firebase')) {
+    transitive = false
+  }
+  compile "com.google.firebase:firebase-core:10.2.6"
```
Comment (HofmannZ) [inline docs/installation-android.md]: Uses version 10.2.6 here, probably wanna bump 😉

File: lib/modules/auth/index.js [hunk 3]
```diff
@@ -157,10 +162,18 @@ export default class Auth extends Base {
   }
 
   /**
+   * Returns a list of authentication providers that can be used to sign in a given user (identified by its main email address).
+   * @return {Promise}
+   */
+  fetchProvidersForEmail(email: string): Promise<Array<String>> {
```
Comment (HofmannZ) [inline lib/modules/auth/index.js]: Thanks a load, needed that! 🙏🏼

**Comment by jasan-s** [comment]
I'm assuming migration to V2 from V1 should be relatively easy and painless. However it would be nice to add a section in docs for migration.

**Comment by mikelambert** [comment]
"we've still got quite a bit to do". Is there a list of bugs, features, etc that block the v2 branch?

This [very-short list](https://github.com/invertase/react-native-firebase/milestone/2) looks to be the closest I can see (as [this list](https://github.com/invertase/react-native-firebase/milestone/1) got pushed out for subsequent releases?) But the v2 list seems too short to believe it represents the "quite a bit to do"... :)

**Comment by Ehesp** [comment]
@mikelambert that was 20 days ago, so we have done quite a bit in that time. We're getting there now though - we should be looking to get v2 out of the door soon, but as there's loads of changes we want to make sure it's as bug free as possible. As for the v2.x list, that was on the v2 list but we decided that we'll never get v2 out if we keep adding to the list hence breaking them out.

**Comment by chrisbianca** [comment]
@jasan-s Yes, it is fairly easy and painless.  At the top of this PR, you'll see a work in progress set of release notes that details any breaking changes that will need to be followed.  We'll apply this to the version release notes when it's released.

**Comment by D1no** [comment]
Hey guys — not cutting short on what great work you guys do: you're awesome! — I'm about to re-implement rn-firebase into our system. First thing will be basic authentication. Should I grab the v2 now to catch the up coming release, or would you advice to go stick with v1?

We are running on react-native v0.45.1 and partially on typescript. So all quite cutting edge.

**Comment by chrisbianca** [comment]
@D1no I'd recommend going straight for v2.  As @ehesp mentions, we're almost there, just trying to make sure it's as stable as possible before an official release.

**Comment by Cretezy** [comment]
Great work! Is v2 fully matched up with the v1 branch as well? If so, we're going to start moving over to v2 and report bugs/PR as we see fit.

**Comment by Salakar** [comment]
@Cretezy v1.x changes have been synced into v2 if that's what you mean?

**Comment by Cretezy** [comment]
Yes, that's what I meant. Thank you!

**Comment by D1no** [comment]
I run into the following issue
image
`/AppDelegate.m:52:24: No known class method for selector 'didReceiveRemoteNotification:'`

 implementing Cloud Messaging according 
https://github.com/invertase/react-native-firebase/blob/v2/docs/installation-ios.md#34-update-appdelegatem

> ### 3.3) Update `AppDelegate.h`
> 
> Add the following import:
> 
> `@import UserNotifications;`
> 
> Change the interface descriptor to:
> 
> `@interface AppDelegate : UIResponder <UIApplicationDelegate,UNUserNotificationCenterDelegate>`
> 
> ### 3.4) Update `AppDelegate.m`
> 
> Add the following import:
> 
> `#import "RNFirebaseMessaging.h"`
> 
> Add the following to the `didFinishLaunchingWithOptions:(NSDictionary *)launchOptions` method after `[FIRApp Configure]`:
> 
> `[[UNUserNotificationCenter currentNotificationCenter] setDelegate:self];`
> 
> Add the following methods:
> 
> ```objectivec
> -(void)application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification {
>   [RNFirebaseMessaging didReceiveLocalNotification:notification];
> }
> 
> - (void)application:(UIApplication *)application didReceiveRemoteNotification:(nonnull NSDictionary *)userInfo {
>   [RNFirebaseMessaging didReceiveRemoteNotification:userInfo];
> }
> 
> - (void)application:(UIApplication *)application didReceiveRemoteNotification:(nonnull NSDictionary *)userInfo
> fetchCompletionHandler:(nonnull void (^)(UIBackgroundFetchResult))completionHandler{
>   [RNFirebaseMessaging didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
> }
> ```


Changing `#import "RNFirebaseMessaging.h"` to `#import <RNFirebaseMessaging.h>` does not lead to a change.

When <kbd>cmd</kbd>+click on `RNFirebaseMessaging.h` brings me to
```objective-c
#ifndef RNFirebaseMessaging_h
#define RNFirebaseMessaging_h
#endif

#import <UIKit/UIKit.h>

#import "Firebase.h"

#if __has_include(<React/RCTEventEmitter.h>)
#import <React/RCTEventEmitter.h>
#else // Compatibility for RN version < 0.40
#import "RCTEventEmitter.h"
#endif
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#else // Compatibility for RN version < 0.40
#import "RCTBridgeModule.h"
#endif
#if __has_include(<React/RCTUtils.h>)
#import <React/RCTUtils.h>
#else // Compatibility for RN version < 0.40
#import "RCTUtils.h"
#endif

@import UserNotifications;

@interface RNFirebaseMessaging : NSObject <RCTBridgeModule>

typedef void (^RCTRemoteNotificationCallback)(UIBackgroundFetchResult result);
typedef void (^RCTWillPresentNotificationCallback)(UNNotificationPresentationOptions result);
typedef void (^RCTNotificationResponseCallback)();

@property (nonatomic, assign) bool connectedToFCM;

#if !TARGET_OS_TV
+ (void)didReceiveRemoteNotification:(nonnull NSDictionary *)userInfo fetchCompletionHandler:(nonnull RCTRemoteNotificationCallback)completionHandler;
+ (void)didReceiveLocalNotification:(nonnull UILocalNotification *)notification;
+ (void)didReceiveNotificationResponse:(nonnull UNNotificationResponse *)response withCompletionHandler:(nonnull RCTNotificationResponseCallback)completionHandler;
+ (void)willPresentNotification:(nonnull UNNotification *)notification withCompletionHandler:(nonnull RCTWillPresentNotificationCallback)completionHandler;
#endif

@end

```

**Comment by Cretezy** [comment]
@D1no have you tried deleting your `node_modules` AND your `yarn.lock`?

**Comment by Salakar** [comment]
Am currently in the process of releasing this so you shouldn't have these probs in a moment.

**Comment by D1no** [comment]
Yes, a clean install. That is the reason I was surprised.

Edit: Awesome @Salakar 😁  than I discard now the yarn via git@ install. I go and get a coffee 🙏

**Comment by Salakar** [comment]
@D1no @Cretezy @mikelambert @jasan-s @HofmannZ @sraka1 v2 has been published 🎉 

A guide to upgrading from v1 to v2 can be found on the docs [here](http://invertase.io/react-native-firebase/#/migration-guide?id=from-v1-to-v2).

**Comment by sraka1** [comment]
Cool! 🎉 

Btw, "confirmPasswordReset" and "applyActionCode" are still missing in Auth. If they aren't added earlier, I will open a PR once I migrate to v2 next week (currently running a fork of the predecessor of this repo).

@Salakar Also, any plans on supporting Dynamic Links in the future? 😄

**Comment by Salakar** [comment]
@sraka1 we'll add those two auth functions to a minor/patch version in the next few days - we wanted to get v2 out the door rather than constantly adding to to todo list and v2 being delayed further.

We've yet to look into links, the next priority feature wise for us is js side firebase app configuration (no more plist/json fiddling) and with that mutli firebase app support.

**Comment by D1no** [comment]
Super Awesome 🚀 ! [Firebase Invites](https://firebase.google.com/docs/invites/) than [App Indexing](https://firebase.google.com/docs/app-indexing/) are certainly super interesting. Dynamic Links is really a question of how that enters the void of react native navigators (we use flux navigator).


> We've yet to look into links, the next priority feature wise for us is js side firebase app configuration (no more plist/json fiddling) and with that mutli firebase app support.

Well, not sure if that is so relevant. All the firebase sdk related material are always surrounded by the plist / json part. Not sure if the abstraction is something worth adding as a new fault line if people start to rely on it.


> - Since we can not use multiple targets, all files are linked to the same xcode target. That can cause issues when including files that should (for simplicity) be of the same reference but vary between builds. Such an example is the `GoogleService-Info.plist` for google login (you could reference them in your AppDelegate file). However, we try to touch as little of the native code as possible to have little hassle in upgrading further rn-versions. The best way to solve such situations is to have an extra folder in your `ios/` folder, which is not added to the xcode project. And have files, depending on their build config suffix be copied to the app at build time. I.e. we manage our firebase files like so:
> manage files
> Here our script (simple shell command):
> ```bash
> cp ${SRCROOT}/firebase/GoogleService-Info-${CONFIGURATION}.plist ${CONFIGURATION_BUILD_DIR}/${UNLOCALIZED_RESOURCES_FOLDER_PATH}/GoogleService-Info.plist
> ```
From: https://github.com/facebook/react-native/issues/11813#issuecomment-310029148

**Comment by Salakar** [comment]
@D1no it's another step in the direction of making it compatible with the web sdk, plus would be a lot less effort required for beginners (something we get a lot of) and additionally it would allow initialising multilpe firebase instances at the same time - just like the web, which has its use cases and has been requested before. Plist/json support will still remain if people want to use it.

**Comment by D1no** [comment]
This was not an intended pull on priorities. Being able to change environments at runtime could be interesting though 🤔 codepush...cough...

**Comment by Salakar** [comment]
@D1no no worries was just stating the reasoning behind it that's all. Flipping projects over via codepush would be cool 😸

**Comment by aarongreenwald** [comment]
@D1no to piggyback on what @Salakar said - I'm one of the people requesting configurability on JS side, although I asked for it after @Salakar wrote that comment :)

I explained my reasoning [in this comment](https://github.com/invertase/react-native-firebase/issues/164#issuecomment-310311240). This isn't just for different environments (although that's a valid use case), it's for runtime production configuration. I don't think code push would even work. So it's definitely a really useful thing to have :)

**Comment by Salakar** [comment]
@sraka1 

> Btw, "confirmPasswordReset" and "applyActionCode" are still missing in Auth. If they aren't added earlier, I will open a PR once I migrate to v2 next week (currently running a fork of the predecessor of this repo).

Implemented as of v2.0.2 - as well as `checkActionCode`.

**Comment by afilp** [comment]
About code-push, I created an issue there:

https://github.com/Microsoft/react-native-code-push/issues/1185#issuecomment-375435954

I believe there can be an "AsyncStorage" usage for applying some settings (like dynamic links) again after code-push is triggered.

What do you think?

**Comment by Salakar** [comment]
@D1no just thought you might like to know as of v4 Firebase Dynamic Links and Firebase Invites are supported: https://github.com/invertase/react-native-firebase/releases/tag/v4.0.0-rc.2 . We've yet to look at App Indexing.

**Pull request merged**